src/module.hpp

```diff
@@ -0,0 +1,236 @@
+    ExportType type;
+    size_t idx;
+
+    Import(std::string module, std::string name, ExportType type, size_t idx)
+        : module(std::move(module)), name(std::move(name)), type(type), idx(idx) {}
+};
+
+class AbstractModule {
+    std::vector<Import> _imports;
+    std::vector<Export> _exports;
+
+    std::vector<AbstractMemory> _memories;
+    std::vector<AbstractFunc> _funcs;
+public:
+    std::vector<Import>& imports() { return _imports; }
+    const std::vector<Import>& imports() const { return _imports; }
+
+    std::vector<Export>& exports() { return _exports; }
+    const std::vector<Export>& exports() const { return _exports; }
+
+    std::vector<AbstractMemory>& memories() { return _memories; }
+    const std::vector<AbstractMemory>& memories() const { return _memories; }
+
+    std::vector<AbstractFunc>& funcs() { return _funcs; }
+    const std::vector<AbstractFunc>& funcs() const { return _funcs; }
+};
+
+class Module {
+    std::vector<Import> _imports;
+    std::vector<Export> _exports;
+
+    std::vector<AbstractMemory> _memories;
+    std::vector<std::shared_ptr<Memory>> _shared_memories;
+
+    std::vector<const FuncSig*> _import_func_sigs;
+    std::vector<std::shared_ptr<UnlinkedFunc>> _funcs;
+
+    Environment* _env;
+public:
+    Module() {}
+    Module(const AbstractModule& abstract, Environment& env);
+
+    Environment& env() const { return *_env; }
+
+    const std::vector<Import>& imports() const { return _imports; }
+    void add_import(Import i) {
+        _imports.emplace_back(i);
+    }
+
+    std::vector<Export>& exports() { return _exports; }
+    void add_export(Export e) {
+        _exports.emplace_back(e);
+    }
+
+    const std::vector<AbstractMemory>& memories() const { return _memories; }
+    void add_memory(AbstractMemory mem) {
+        if (mem.is_shared && !mem.is_import) {
+            _shared_memories.emplace_back(new Memory(mem));
+        } else {
+            _shared_memories.emplace_back(nullptr);
+        }
+
+        _memories.emplace_back(mem);
+    }
+
+    const std::vector<std::shared_ptr<UnlinkedFunc>>& funcs() const { return _funcs; }
+    void add_func(std::shared_ptr<UnlinkedFunc> func) {
+        _import_func_sigs.emplace_back(nullptr);
+        _funcs.emplace_back(func);
+    }
+    void add_imported_func(FuncSig* sig) {
+        _import_func_sigs.emplace_back(sig);
+        _funcs.emplace_back(nullptr);
+    }
+
+    friend class ModuleInstance;
+};
+
+class ImportModule {
+public:
+    virtual LinkedFunc* find_func(const Import& import) const = 0;
+    virtual std::shared_ptr<Memory> find_memory(const Import& import) const = 0;
+};
+
+class ImportMultiModule : public ImportModule {
+    std::vector<ImportModule*> _modules;
+public:
+    ImportMultiModule(std::vector<ImportModule*>&& modules) : _modules(modules) {}
+
+    LinkedFunc* find_func(const Import& import) const override;
+    std::shared_ptr<Memory> find_memory(const Import& import) const override;
```

Comment:
Yes. While we don't actually modify the module we're linking to (in the sense that we haven't added or removed anything from it), the returned values definitely need to be non-`const`, since they're going to be linked into code which can be executed to modify them later.

Reply:
Got it.